components/clp-py-utils/clp_py_utils/s3_utils.py

```diff
@@ -1,11 +1,137 @@
+    """
+
+    aws_access_key_id = None
+    aws_secret_access_key = None
+
+    if not credentials_file_path.exists():
+        raise ValueError(f"File {credentials_file_path} doesn't exist.")
+
+    with open(credentials_file_path, "r") as f:
+        for line in f:
+            line = line.strip()
+            if line.startswith("aws_access_key_id"):
+                aws_access_key_id = line.split("=", 1)[1].strip()
+            elif line.startswith("aws_secret_access_key"):
+                aws_secret_access_key = line.split("=", 1)[1].strip()
+
+    if aws_access_key_id is None or aws_secret_access_key is None:
+        raise ValueError(
+            "The credentials file must contain both 'aws_access_key_id' and 'aws_secret_access_key'."
+        )
+
+    return aws_access_key_id, aws_secret_access_key
+
+
+def parse_s3_url(s3_url: str) -> Tuple[str, str, str]:
+    """
+    Parses the region_code, bucket and key_prefix from the given S3 url. The url must be either a
+    host_style_url or path_style_url.
+    :param s3_url: a host_style_url or path_style_url.
+    :return: A tuple of (region_code, bucket, key_prefix)
+    :raise: ValueError if the given s3_url is not a valid host_style_url or path_style_url.
+    """
+
+    host_style_url_regex = re.compile(
+        r"https://(?P<bucket_name>[a-z0-9.-]+)\.s3(\.(?P<region_code>[a-z0-9-]+))?"
+        r"\.(?P<endpoint>[a-z0-9.-]+)/(?P<key_prefix>[^?]+).*"
+    )
+    match = host_style_url_regex.match(s3_url)
+
+    if match is None:
+        path_style_url_regex = re.compile(
+            r"https://s3(\.(?P<region_code>[a-z0-9-]+))?\.(?P<endpoint>[a-z0-9.-]+)/"
+            r"(?P<bucket_name>[a-z0-9.-]+)/(?P<key_prefix>[^?]+).*"
+        )
+        match = path_style_url_regex.match(s3_url)
+
+    if match is None:
+        raise ValueError(f"Unsupported URL format: {s3_url}")
+
+    region_code = match.group("region_code")
+    bucket_name = match.group("bucket_name")
+    endpoint = match.group("endpoint")
+    key_prefix = match.group("key_prefix")
+
+    if AWS_ENDPOINT != endpoint:
+        raise ValueError(f"Unsupported endpoint: {endpoint}")
+
+    return region_code, bucket_name, key_prefix
+
+
+def generate_s3_virtual_hosted_style_url(
+    region_code: str, bucket_name: str, object_key: str
+) -> str:
+    return f"https://{bucket_name}.s3.{region_code}.{AWS_ENDPOINT}/{object_key}"
+
+
+def get_s3_object_metadata(s3_input_config: S3InputConfig) -> Result[List[FileMetadata], str]:
+    """
+    Gets the metadata of objects under the <bucket>/<key_prefix> specified by s3_input_config.
+    Note: We reuse FileMetadata class to store the metadata of S3 objects. The object_key is stored
+    as path in FileMetadata.
+
+    :param s3_input_config: S3 configuration specifying the bucket, key_prefix and credentials.
+    :return: Result.OK(List[FileMetadata]) containing the object metadata on success,
+             otherwise Result.Err(str) with the error message.
```

Comment:
```suggestion
    :param s3_input_config:
    :return: Result.OK(List[FileMetadata]) containing the object's metadata on success,
    otherwise Result.Err(str) with the error message.
```